core/src/main/scala/chisel3/Data.scala

```diff
@@ -661,10 +661,33 @@ abstract class Data extends HasId with NamedComponent with SourceInfoDoc {
           )
       }
     } else {
-      this.legacyConnect(that)
+      this.firrtlPartialConnect(that)
     }
   }
 
+  private[chisel3] def directionalBulkConnect(
+    that: Data
+  )(
+    implicit sourceInfo:   SourceInfo,
+    connectCompileOptions: CompileOptions
+  ): Unit = {
+    if (connectCompileOptions.checkSynthesizable) {
+      requireIsHardware(this, s"LHS (sink) data to be bulk-connected")
+      requireIsHardware(that, s"RHS (source) data to be bulk-connected")
+      (this.topBinding, that.topBinding) match {
+        case (_: ReadOnlyBinding, _: ReadOnlyBinding) => throwException(s"Both $this and $that are read-only")
+        // DontCare cannot be a sink (LHS)
+        case (_: DontCareBinding, _) => throw BiConnect.DontCareCantBeSink
+        case _ => // fine
+      }
+    }
+    if (!BiConnect.canFirrtlConnectData(this, that, sourceInfo, connectCompileOptions, Builder.referenceUserModule)) {
+      Builder.error(s"Cannot issue a directional bulk connect $this :<>= $that")
+    }
+    // TODO write error-checking logic, so it doesn't require the FIRRTL error if `this` is not writable
```

Comment:
Hasn't this `TODO` been taken care of by the change of Node(this) to lref inside this function? Also doesn't canFirrtlConnectData have a bunch of the error checking? I am deleting this TODO
```suggestion
```